db/commitlog/commitlog.cc

```diff
@@ -1779,6 +1780,76 @@ void db::commitlog::segment_manager::flush_segments(uint64_t size_to_remove) {
     totals.bytes_flush_requested += flushing;
 }
 
+void db::commitlog::segment_manager::check_no_data_older_than_allowed() {
+    if (!cfg.commitlog_data_max_lifetime_in_seconds) {
+        return;
+    }
+
+    auto now = gc_clock::now();
+
+    std::unordered_set<cf_id_type> ids;
+    std::optional<replay_position> high;
+
+    for (auto& s : _segments) {
+        auto rp = replay_position(s->_desc.id, db::position_type(s->size_on_disk()));
+        if (rp <= _flush_position) {
+            // already requested.
+            continue;
```

Comment:
Wouldn't it be better to have a per-segment timer? Ideally, the timer never expires.

The number of segments we have is ~ 8GB/32MB = 256. Each segment could reference many memtables. So potentially a long loop.

Reply:
? This has nothing to do with timer. We are just filtering out segments that are already flush-requested for.

Reply:
And looping 256 objects is hardly a very long loop. Computation per segment is low, not to mention a lot fall inside filter and is not even considered.

Reply:
Talking about the outer loop, around _segments, then the inner loop around _cf_min_time.

Let's say we have 100 old segments, each referencing 100 tables (or 100 tablets). Won't we issue 10,000 ops?

Reply:
And this is called from on_timer. So there is amplification here, each timer expiration we loop on all segments.

Reply:
is this still a problem? is it acceptable?

Reply:
Note that the amplification you mention can't really happen unless you either have super short max time or very poor flushing. We stop the loop as soon as we find a segment with no data older than max (which will most of the time be the first segment). Since data is inserted and processed in order, we never need to process further than first segment that both has data left and has no data that should be expired. Similarly, anything already requested, we skip. End result is that it is a fairly small window that is processed on any given call.

Reply:
Ok